### ISSUE TITLE
Cache ssh connections

### ISSUE THREAD

**Issue by elyezer** [SatelliteQE/robottelo#2678]
By reusing the same connection we can speed up up to 4 times:

``` python
import paramiko
import timeit


CONNECTIONS = []


def get_client():
    client = paramiko.SSHClient()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    return client


def client_connect(client):
    client.connect(
        hostname='localhost',
        username='user',
        password='password'
    )


def command_without_cached_connection(cmd):
    client = get_client()
    client_connect(client)
    _, stdout, stderr = client.exec_command(cmd)
    exit_status = stdout.channel.recv_exit_status()
    stdout = stdout.read()
    stderr = stderr.read()


def command_with_cached_connection(cmd):
    if len(CONNECTIONS) == 0:
        client = get_client()
        client_connect(client)
        CONNECTIONS.append(client)
    else:
        client = CONNECTIONS[0]
    _, stdout, stderr = client.exec_command(cmd)
    exit_status = stdout.channel.recv_exit_status()
    stdout = stdout.read()
    stderr = stderr.read()


if __name__ == '__main__':
    print('Cached:')
    print(timeit.timeit(
        'command_with_cached_connection("ls")',
        setup='from __main__ import command_with_cached_connection',
        number=10
    ))

    print('Not cached:')
    print(timeit.timeit(
        'command_without_cached_connection("ls")',
        setup='from __main__ import command_without_cached_connection',
        number=10
    ))
```

Running on my machine I get:

```
Cached:
1.52692103386
Not cached:
6.2257540226
```

**Comment by sthirugn** [comment]
wow!

**Comment by rplevka** [comment]
:+1: 
we could reuse the `common.ssh._get_connection()`.
but it is not returning a `SSHClient` as the docstring says, but it rather yelds a generator.
so before using it, one needs to call `__enter__()`.

```python
>>> from robottelo.common import ssh
>>> b = a.__enter__()
>>> stdin, stdout, stderr = b.exec_command('touch foo')
>>> stdout.channel.exit_status
0
>>> stdin, stdout, stderr = b.exec_command('ls -l foo')
>>> stdout.read()
'-rw-r--r--. 1 root root 0 Aug 13 05:57 foo\n'
>>>
```

let's implement it!

**Comment by Ichimonji10** [comment]
> By reusing the same connection we can speed up up to 4 times:

Do our profiling tools agree? IIRC, SSH connections accounted for a small portion of the SSH test suite time.

**Comment by Ichimonji10** [comment]
I checked out the most recent results for `satellite6-automation-downstream-rhel67-cli`. SSH connections accounted for ~5% of the test suite execution time. So, yes, there's benefit to caching SSH connections. But it's a ~5% benefit, not a 400% benefit. I'm all for improving Robottelo, but let's be honest with ourselves here.

**Comment by rplevka** [comment]
@Ichimonji10, looking at @elyezer 's example, I believe he's referring to the execution of the ssh commands, not entire automation as you do. Simple as that.

**Comment by Ichimonji10** [comment]
Understood. But what's the point of looking at just ssh command execution? It's neat, but it's not very relevant to Robottelo.

**Comment by elyezer** [comment]
The relevance here is that a customer would ssh into the server and execute some scenario using the same ssh session, not creating one for every single command. This will mimic more the customer behaviour and will speed up how connections are managed.

Also the idea is to try to make the connection on setUpClass and tearDownClass. This is not something major though but have its value.

**Comment by Ichimonji10** [comment]
> This will mimic more the customer behaviour and will speed up how connections are managed.

Ahhh. I hadn't thought about that. It would be nice if Robottelo acts more like a customer.

I clarified internally with @elyezer, but I want to clarify here too: I do like the idea of caching SSH connections. Please don't take my comments too harshly. I'm just trying to make sure our rationale for making changes is sound.

**Comment by elyezer** [comment]
The connections should have thread safety access because the performance tests.

**Comment by renzon** [comment]
@rochacbruno take look at this once we discussed about it. It seems this is a cyclical conversation ;)

**Comment by rochacbruno** [comment]
I opened https://github.com/SatelliteQE/robottelo/issues/3694 to adress this.

Also we already have a function `ssh.execute_command` which expects a connection to reuse.

**Comment by rochacbruno** [comment]
BTW: @omaciel mentioned that we should in future replace paramiko SSH with the tooling @elyezer and @Ichimonji10 are using on Pulp.

**Comment by omaciel** [comment]
@rochacbruno thanks for remembering it :)

**Comment by Ichimonji10** [comment]
Pulp Smash has a class, `cli.Client`, that executes commands in one of two ways:

* Commands may be executed in a local shell. This is useful when the application under test (Pulp for us, Satellite for you) is running on the local system.
* Commands may be executed via SSH. This is useful when the application under test is running on a remote system.

When commands are executed via SSH, the system-wide `ssh` executable is used to execute commands. I'd like to make us less dependent on OpenSSH in particular, so that we could use e.g. the dropbear SSH client, but for now, calls to `ssh` are hard-coded in. OpenSSH offers connection caching, and the necessary options can be specified either on the command line when running a command, or in `~/.ssh/config`. Pulp Smash currently expects the user to set necessary options in the config file - not totally user friendly, but so it is. Here's what the `~/.ssh/config` options might look like:

```
Host foo.example.com bar.example.com
    User root
    IdentityFile ~/.ssh/common-private-key.example.com
    StrictHostKeyChecking no
    UserKnownHostsFile /dev/null

    ControlMaster auto
    ControlPersist 5m
    ControlPath ~/.ssh/controlmasters/%C
```

This configuration assumes that a `~/.ssh/controlmasters` directory exists. If it doesn't, you'll encounter some weird error. See `ssh_config(5)` for detailed information on the options I've pasted above. The `Control*` options are the relevant ones.

**Comment by elyezer** [comment]
Using the same client as Pulp Smash will help also the release process team since they need to run robottelo locally on the machine they provision, executing commands locally instead of running ssh for localhost is better in that case.

Also removing the SSH management from robottelo the environment can be provisioned and then robottelo just need to know the hostname, nothing more.

**Comment by rochacbruno** [comment]
Fixed in #3943 

```
with ssh.get_connection() as connection:
    connection.run('command')
    ...
```
